lib/src/position.dart

```diff
@@ -1722,6 +1723,149 @@ class ThreeCheck extends Position<ThreeCheck> {
   }
 }
 
+/// A variant where the goal is to put your king on the eigth rank
+@immutable
+class RacingKings extends Position<RacingKings> {
+  const RacingKings({
+    required super.board,
+    super.pockets,
+    required super.turn,
+    required super.castles,
+    super.epSquare,
+    required super.halfmoves,
+    required super.fullmoves,
+  });
+
+  const RacingKings._initial()
+      : super(
+            board: Board.racingKings,
+            pockets: null,
+            turn: Side.white,
+            castles: Castles.empty,
+            epSquare: null,
+            halfmoves: 0,
+            fullmoves: 1);
+
+  static const initial = RacingKings._initial();
+  static const goal = SquareSet.fromRank(7);
+
+  bool get blackCanReachGoal {
+    final blackKing = board.kingOf(Side.black);
+    return blackKing != null &&
+        kingAttacks(blackKing).intersect(goal).squares.where((square) {
+          // Check whether this king move is legal
+          final context = _Context(
+            isVariantEnd: false,
+            mustCapture: false,
+            king: blackKing,
+            blockers: _sliderBlockers(blackKing),
+            checkers: checkers,
+          );
+          final legalMoves = _legalMovesOf(blackKing, context: context);
+          return legalMoves.has(square);
+        }).isNotEmpty;
+  }
+
+  bool get blackInGoal =>
+      board.black.intersect(goal).intersect(board.kings).isNotEmpty;
+  bool get whiteInGoal =>
+      board.white.intersect(goal).intersect(board.kings).isNotEmpty;
+
+  @override
+  SquareSet _legalMovesOf(Square square, {_Context? context}) =>
+      SquareSet.fromSquares(super
+          ._legalMovesOf(square, context: context)
+          .squares
+          .where((to) =>
+              !playUnchecked(NormalMove(from: square, to: to)).isCheck));
+
+  @override
+  bool isLegal(Move move) =>
+      !playUnchecked(move).isCheck && super.isLegal(move);
+
+  @override
+  bool get isVariantEnd {
+    if (!whiteInGoal && !blackInGoal) {
+      return false;
+    }
+    if (blackInGoal || !blackCanReachGoal) {
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  Outcome? get variantOutcome {
+    if (!isVariantEnd) return null;
+    if (whiteInGoal && blackInGoal) return Outcome.draw;
+    // If white is in the goal, check
+    // whether black can reach the goal
+    // And vice-versa
```

Comment:
This comment is not quite correct. We only do this check if White is in goal, not the other way around.

Reply:
Resolved in b637550